### PR TITLE
Add server-side validation of submissions

### LINKED ISSUE #758: Server-side validation of Responses
**Describe the bug**
Requirements such as making a specific question mandatory seem to only be enforced client-side (e.g. in the user's browser).  If a user intercepts the HTTP request with their answer and changes the drop-down index to a non-existent one, the form response is accepted by the server and the question is marked as "No Response" (when that should not be possible as the question was marked as mandatory)

**To Reproduce**
1. Create a Form.
2. Add a drop-down question, marked as 'Required'
2. Go to the form link and select one of the answers in the drop-down.
3. Intercept the request to the server (e.g. using a browser extension such as TamperData or an intercepting proxy such as ZAP or Burp)
4. Replace the index of the answer by a non-existent one (e.g. if the answer had 3 options in the drop-down, change the selected index to '5', for example).
5. The response will be accepted.
6. When the responses are viewed by the Form creator, the drop-down answers that were tampered with non-existent indices will appear as "No response".

**Expected behavior**
The server should reject the response informing the user that a non-valid answer to the question was chosen/provided.

**Nextcloud:**
- Nextcloud-Version: 20.0.6
- Forms-Version: 2.1.0

### PR DESCRIPTION
**Pull request by Chartman123**
Fixes #758 by checking answers before creating a submission and sending an error "400 - Bad Request" if at least one question has no valid answer. 

- [x] required questions have at least one answer
- [x] answers are in options
- [x] questions don't have multiple answers if only one answer is expected
- [x] ~~valid date / datetime values~~ (wait for #736)
- [x] all answered questions correspond to the current form

Signed-off-by: Christian Hartmann <chris-hartmann@gmx.de>

### CODE REVIEW

**Comment by Chartman123** [comment]
@jotoeri Any other cases that should get validated? If so, just add the items to the list in the first comment 👍🏻

**Comment by jotoeri** [comment]
I added one more thing, that currently came to my mind... 👍

**Comment by Chartman123** [comment]
I think that this will already be filtered in the insertion itself, but we can of course add this to the validation, too :)

**Comment by jotoeri** [comment]
Currently yes, but we can change that... 😉 
So just from logical side - if we now have a validation in front of the insertion, that checks if the options are valid, then the logical way is to also check the questions there, having the same handling if invalid. We can even simplify the actual insertion then, as we know already, that when called it's valid.

**Comment by Chartman123** [comment]
Are you aware of an easy way to parse the moment.js date/datetime strings in PHP? As far as I've seen, the values are transmitted localized as a string and are also stored like that inside the DB. Will it work to feed these strings to the date_parse() function?

**Comment by jotoeri** [comment]
Hmm, honestly no. But we might also have to change that format somhow again. (See e.g. #736 ) So don't put too much effort into checking this. The problem here is especially, that it might have to fit all localized formats currently. Including especially all languages.

**Comment by Chartman123** [comment]
Ok, I came up with an solution that might probably work... But I think that this might need some testing for several date formats... Otherwise I think that I have covered all validations now :)

**Comment by Chartman123** [comment]
Ok, seems like I still have some errors in my code. I tried to send a submission from within the app (not with RESTED) and got a 400 from my validation 😦

**Comment by Chartman123** [comment]
> Ok, seems like I still have some errors in my code. I tried to send a submission from within the app (not with RESTED) and got a 400 from my validation 😦

Should be fixed by my latest commit 👍🏻

**Comment by jotoeri** [comment]
Hmm, so probably we should already add the tests before merging then... Especially on different languages for dates, so we don't start blocking any users now...
Do you want to have a look into? Or should i give it a go?

**Comment by Chartman123** [comment]
You mean automated unit tests? I have never done that before. So I wouldn't mind if you did it :)

**Comment by jotoeri** [comment]
> We could also just leave the date validation out of the PR for now and wait for a standardized date format for the storage.

That's probably a better way.... It's just quite error-prone, so even if we test 10 Languages, there might be another one, where it fails...

**Comment by jotoeri** [comment]
Ahh, now i was just a second too slow... 😄 
Can you rebase onto current master? So just git pull on master and then rebase the branch... This just as a general thing for a clean history... 😉

**Comment by Chartman123** [comment]
> Can you rebase onto current master? So just git pull on master and then rebase the branch...

Better now?

**Comment by jotoeri** [comment]
Yep 👍 
grafik

**Comment by jotoeri** [comment]
Just added some tests... This one was unexpectedly easy, as it does not call anything else than default php-functions 😄 🙈 

So if you still have ideas, what to test, you can just append them to the dataProvider...

**Comment by Chartman123** [comment]
That's basically what I tested manually during development, so I think we're good to go now 👍🏻

File: lib/Service/SubmissionService.php [hunk 1]
```diff
@@ -240,4 +241,58 @@ private function array2csv(array $header, array $records): string {
...
+						if (array_search($answer, array_column($question['options'], 'id')) === false) {
+							return false;
+						}
+					}
+				}
+	
+				// TODO Check if date questions have valid answers
+				if (in_array($question['type'], [Constants::ANSWER_TYPE_DATE, Constants::ANSWER_TYPE_DATETIME])) {
+					continue;
+				}
```
Comment (jotoeri) [inline lib/Service/SubmissionService.php:283]: Hmm, forgot to add before commiting or so? 😉

Comment (Chartman123) [reply lib/Service/SubmissionService.php:283]: I wanted to leave it here as a TODO for the future. Should I remove it completely?

Comment (jotoeri) [reply lib/Service/SubmissionService.php:283]: Ahh, now i see, you just removed the return...^^
Yes, better remove, it's just dead code like this....

Comment (Chartman123) [reply lib/Service/SubmissionService.php:283]: Exactly... So remove or put it in comments?

**Review by jotoeri — APPROVED** [review]

**Comment by jancborchardt** [comment]
Awesome contribution @Chartman123! We have a Nextcloud Talk instance for contributors, with e.g. a general "Community chat" and also a channel for the Forms app for quicker coordination. If you are interested, I can add you there with a guest account via your email address. :)

**Comment by Chartman123** [comment]
@jancborchardt thanks 😇 yes, that would be nice. 👍🏻

**Pull request merged**